src/VisualStudio/Core/Def/Implementation/Experimentation/AnalyzerVsixSuggestedActionCallback.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.Editor.Implementation.Suggestions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Experimentation
+{
+    [Export(typeof(ISuggestedActionCallback))]
+    internal class AnalyzerVsixSuggestedActionCallback : ForegroundThreadAffinitizedObject, ISuggestedActionCallback
+    {
+        private const string AnalyzerEnabledFlight = @"LiveCA/LiveCAcf";
+        private const string AnalyzerVsixHyperlink = @"https://go.microsoft.com/fwlink/?linkid=849061";
+        private static readonly Guid FxCopAnalyzersPackageGuid = Guid.Parse("{4A41D270-A97F-4639-A352-28732FC410E4}");
+
+        private readonly VisualStudioWorkspace _workspace;
+        private readonly SVsServiceProvider _serviceProvider;
+
+        /// <summary>
+        /// Tracks when the bar is shown so we don't show to the user more than once per session
+        /// </summary>
+        private bool _infoBarShown = false;
+
+        /// <summary>
+        /// This service is initialzed by <see cref="OnSuggestedActionExecuted(SuggestedAction)"/>
+        /// </summary>
+        private IExperimentationService _experimentationService;
+
+        private LiveCAInstallStatus _installStatus = LiveCAInstallStatus.Unchecked;
+
+        [ImportingConstructor]
+        public AnalyzerVsixSuggestedActionCallback(VisualStudioWorkspace workspace,
+            SVsServiceProvider serviceProvider)
+        {
+            _workspace = workspace;
+            _serviceProvider = serviceProvider;
+        }
+
+        public void OnSuggestedActionExecuted(SuggestedAction action)
+        {
+            // If the experimentation service hasn't been retrieved, we'll need to be on the UI
+            // thread to get it
+            AssertIsForeground();
+
+            // Initialize the experimentation service if it hasn't yet been fetched
+            if (_experimentationService == null)
+            {
+                _experimentationService = _workspace.Services.GetRequiredService<IExperimentationService>();
+            }
+
+            // If we haven't yet checked to see if the VSIX is installed, check now
+            if (_installStatus == LiveCAInstallStatus.Unchecked)
```

Comment:
extrac ll of this into a helper.  Then you won't need all this complex control flow.